primary/src/grpc_server/validator.rs

```diff
@@ -14,35 +19,89 @@ use tonic::{Request, Response, Status};
 use types::{
     BatchMessageProto, BlockError, BlockRemoverErrorKind, CertificateDigest,
     CertificateDigestProto, CollectionRetrievalResult, Empty, GetCollectionsRequest,
-    GetCollectionsResponse, RemoveCollectionsRequest, Validator,
+    GetCollectionsResponse, ReadCausalRequest, ReadCausalResponse, RemoveCollectionsRequest,
+    Validator,
 };
 
-#[derive(Debug)]
-pub struct NarwhalValidator {
+pub struct NarwhalValidator<
+    PublicKey: VerifyingKey,
+    SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static,
+> {
     tx_get_block_commands: Sender<BlockCommand>,
     tx_block_removal_commands: Sender<BlockRemoverCommand>,
     get_collections_timeout: Duration,
     remove_collections_timeout: Duration,
+    block_synchronizer_handler: Arc<SynchronizerHandler>,
+    dag: Option<Arc<Dag<PublicKey>>>,
 }
 
-impl NarwhalValidator {
+impl<PublicKey: VerifyingKey, SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static>
+    NarwhalValidator<PublicKey, SynchronizerHandler>
+{
     pub fn new(
         tx_get_block_commands: Sender<BlockCommand>,
         tx_block_removal_commands: Sender<BlockRemoverCommand>,
         get_collections_timeout: Duration,
         remove_collections_timeout: Duration,
+        block_synchronizer_handler: Arc<SynchronizerHandler>,
+        dag: Option<Arc<Dag<PublicKey>>>,
     ) -> Self {
         Self {
             tx_get_block_commands,
             tx_block_removal_commands,
             get_collections_timeout,
             remove_collections_timeout,
+            block_synchronizer_handler,
+            dag,
         }
     }
 }
 
 #[tonic::async_trait]
-impl Validator for NarwhalValidator {
+impl<PublicKey: VerifyingKey, SynchronizerHandler: Handler<PublicKey> + Send + Sync + 'static>
+    Validator for NarwhalValidator<PublicKey, SynchronizerHandler>
+{
+    async fn read_causal(
+        &self,
+        request: Request<ReadCausalRequest>,
+    ) -> Result<Response<ReadCausalResponse>, Status> {
+        let collection_id = request
+            .into_inner()
+            .collection_id
+            .ok_or_else(|| Status::invalid_argument("Missing collection"))?;
+        let ids = parse_certificate_digests(vec![collection_id])?;
+
+        let block_header_results = self
+            .block_synchronizer_handler
+            .get_and_synchronize_block_headers(ids.clone())
+            .await;
+
+        for result in block_header_results {
+            if result.is_err() {
+                // TODO: Should we capture & return each individual sync error?
+                return Err(Status::internal(format!(
```

Comment:
I would say returning just one error for now is ok. 

You can also write the statement as:

```
if let Err(err) = result {
    return Err(Status::internal(format!("Error when trying to synchronize block headers: {:?}", err)));
}
```

so you don't have to deal with `unwrap_err` etc